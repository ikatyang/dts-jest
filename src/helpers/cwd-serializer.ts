export const cwd_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(new RegExp(process.cwd(), 'g'), '<cwd>')),
  test: (value: any) =>
    typeof value === 'string' && value.indexOf(process.cwd()) !== -1,
};
