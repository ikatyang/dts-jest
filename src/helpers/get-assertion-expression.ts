import {
  runtime_namespace,
  IRuntimeReportParameters,
  IRuntimeSnapshotParameters,
  TriggerKind,
} from '../definitions';

export const get_assertion_expression = (kind: TriggerKind, line: number) => {
  if (kind === TriggerKind.Show) {
    const report_parameters: IRuntimeReportParameters = {line};
    const report_value = `${runtime_namespace}.report(${JSON.stringify(report_parameters)})`;
    return `console.log(${report_value})`;
  }
  const snapshot_parameters: IRuntimeSnapshotParameters = {line};
  const snapshot_value = `${runtime_namespace}.snapshot(${JSON.stringify(snapshot_parameters)})`;
  return `expect(${snapshot_value}).toMatchSnapshot()`;
};
