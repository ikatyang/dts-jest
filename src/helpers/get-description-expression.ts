import {
  runtime_namespace,
  IRuntimeDescriptionParameters,
} from '../definitions';

export const get_description_expression = (line: number) => {
  const parameters: IRuntimeDescriptionParameters = {
    line,
  };
  return `${runtime_namespace}.description(${JSON.stringify(parameters)})`;
};
