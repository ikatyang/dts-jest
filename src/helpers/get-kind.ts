import {TriggerKind} from '../definitions';

export const get_kind = (flag?: string) => {
  switch (flag) {
    case ':show':
      return TriggerKind.Show;
    case ':skip':
      return TriggerKind.Skip;
    case ':only':
      return TriggerKind.Only;
    case undefined:
      return TriggerKind.None;
    default:
      throw new Error(`Unexpected flag ${JSON.stringify(flag)}`);
  }
};
