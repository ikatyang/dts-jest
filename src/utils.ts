import * as request from 'request';
import * as ts from 'typescript';
import {ServerPage} from './server';

export const traverse_node = (node: ts.Node, callback: (node: ts.Node) => void) => {
  if (node.kind !== ts.SyntaxKind.SourceFile) {
    callback(node);
  }
  ts.forEachChild(node, child => traverse_node(child, callback));
};

export const remove_spaces = (text: string) =>
  text.replace(/^\s+|\s+$|\n/mg, '');

export const defaults = <T>(value: undefined | T, default_value: T): T =>
  (value === undefined)
    ? default_value
    : value;

export const request_server = (port: number, page: ServerPage, args?: {}, callback?: (body: string) => void) => {
  const url = `http://127.0.0.1:${port}${page}`;
  return request.get(
    url,
    {
      qs: args,
    },
    (error, response, body) => {
      // istanbul ignore next
      if (error || response.statusCode !== 200) {
        throw new Error(`Error: url=${url} args=${args} -> ${
          error
            ? error
            : `status=${response.statusCode}`
        }`);
      } else if (callback) {
        callback(body);
      }
    },
  );
};
