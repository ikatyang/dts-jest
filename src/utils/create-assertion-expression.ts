import { Trigger, TriggerFooterFlag, TriggerHeaderFlags } from '../definitions';

export interface CreateAssertionExpressionOptions {
  test_type: boolean;
  test_value: boolean;
  get_type_inference_or_diagnostic_expression: string;
  get_type_inference_or_throw_diagnostic_expression: string;
  get_type_report_expression: string;
  get_value_report_expression: string;
}

export const snapshot_assertion_message = '(type) should match snapshot';

export const create_assertion_expression = (
  trigger: Trigger,
  options: CreateAssertionExpressionOptions,
) => {
  const expressions: string[] = [];

  const { header, body, footer } = trigger;

  push_type_show_if_available();
  push_type_pass_if_available();
  push_type_fail_if_available();
  push_type_snap_if_available();
  push_type_not_any_if_available();
  push_value_show_if_available();
  push_value_equal_if_available();
  push_value_error_if_available();
  push_value_no_error_if_available();

  return expressions.join(';');

  function push_type_show_if_available() {
    if (options.test_type && header.flags & TriggerHeaderFlags[':show']) {
      expressions.push(
        create_wrapper(
          '(type) should show report',
          `console.log(${options.get_type_report_expression})`,
        ),
      );
    }
  }

  function push_type_pass_if_available() {
    if (options.test_type && header.flags & TriggerHeaderFlags[':pass']) {
      expressions.push(
        create_wrapper(
          '(type) should not throw error',
          `expect(function () { ${options.get_type_inference_or_throw_diagnostic_expression} }).not.toThrowError()`,
        ),
      );
    }
  }

  function push_type_fail_if_available() {
    if (options.test_type && header.flags & TriggerHeaderFlags[':fail']) {
      expressions.push(
        create_wrapper(
          '(type) should throw error',
          `expect(function () { ${options.get_type_inference_or_throw_diagnostic_expression} }).toThrowError()`,
        ),
      );
    }
  }

  function push_type_snap_if_available() {
    if (options.test_type && header.flags & TriggerHeaderFlags[':snap']) {
      expressions.push(
        create_wrapper(
          snapshot_assertion_message,
          `expect(${options.get_type_inference_or_diagnostic_expression}).toMatchSnapshot()`,
        ),
      );
    }
  }

  function push_type_not_any_if_available() {
    if (options.test_type && header.flags & TriggerHeaderFlags[':not-any']) {
      expressions.push(
        create_wrapper(
          '(type) should not be any',
          `expect(${options.get_type_inference_or_diagnostic_expression}).not.toBe("any")`,
        ),
      );
    }
  }

  function push_value_equal_if_available() {
    if (
      options.test_value &&
      footer !== undefined &&
      footer.flag === undefined
    ) {
      expressions.push(
        create_wrapper(
          `(value) should equal to ${footer.expected}`,
          `expect(${body.experssion}).toEqual(${footer.expected})`,
        ),
      );
    }
  }

  function push_value_error_if_available() {
    if (
      options.test_value &&
      footer !== undefined &&
      footer.flag === TriggerFooterFlag.Error
    ) {
      expressions.push(
        create_wrapper(
          '(value) should throw error',
          `expect(function () { ${body.experssion} }).toThrowError()`,
        ),
      );
    }
  }

  function push_value_no_error_if_available() {
    if (
      options.test_value &&
      footer !== undefined &&
      footer.flag === TriggerFooterFlag.NoError
    ) {
      expressions.push(
        create_wrapper(
          '(value) should not throw error',
          `expect(function () { ${body.experssion} }).not.toThrowError()`,
        ),
      );
    }
  }

  function push_value_show_if_available() {
    if (
      options.test_value &&
      footer !== undefined &&
      footer.flag === TriggerFooterFlag.Show
    ) {
      expressions.push(
        create_wrapper(
          '(value) should show report',
          `console.log(${options.get_value_report_expression})`,
        ),
      );
    }
  }

  function create_wrapper(title: string, expression: string) {
    const description = JSON.stringify(title);
    const {
      header: { method },
    } = trigger;
    return `${method}(${description}, function () { ${expression} })`;
  }
};
