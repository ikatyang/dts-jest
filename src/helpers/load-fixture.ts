import * as fs from 'fs';
import * as path from 'path';
import * as _ts from 'typescript';
import { create_source_file } from '../utils/create-source-file';

export const get_fixture_filename = (id: string) =>
  path.resolve(__dirname, '../../fixtures', id);

export const load_fixture = (id: string) =>
  fs.readFileSync(get_fixture_filename(id), 'utf8');

export const load_fixture_source_file = (id: string, ts: typeof _ts) => {
  const source = load_fixture(id);
  return create_source_file(id, source, ts);
};
