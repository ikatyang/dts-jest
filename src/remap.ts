import * as jest_snapshot_parser from 'jest-snapshot-parser';
import * as _ts from 'typescript';
import { AssertionFlag, TestMethod, Trigger } from './definitions';
import { create_triggers } from './utils/create-triggers';
import { default_to } from './utils/default-to';
import { get_formatted_description } from './utils/get-formatted-description';
import { skipify_triggers_method } from './utils/skipify-triggers-method';

// tslint:disable-next-line:no-duplicate-imports no-unused-variable
import { Parsed } from 'jest-snapshot-parser';

export interface RemapOptions {
  typescript?: typeof _ts;
  source_filename?: string;
}

export const remap = (
  source_content: string,
  snapshot_content: string | jest_snapshot_parser.Parsed,
  // istanbul ignore next
  options: RemapOptions = {},
) => {
  const { typescript: ts = _ts, source_filename = '' } = options;

  const parsed_snapshot =
    typeof snapshot_content === 'string'
      ? jest_snapshot_parser.parse(snapshot_content)
      : snapshot_content;

  const source_file = ts.createSourceFile(
    source_filename,
    source_content,
    ts.ScriptTarget.Latest,
    false,
  );

  const triggers = create_triggers(source_file, ts);
  skipify_triggers_method(triggers);

  const source_line_contents = source_content.split('\n');

  const counters: Record<string, number> = {};
  triggers
    .filter(
      trigger =>
        trigger.flags.indexOf(AssertionFlag.Snapshot) !== -1 &&
        trigger.method !== TestMethod.Skip,
    )
    .forEach(trigger => {
      const base_title = get_snapshot_base_title(trigger);
      const counter = get_increased_counter(base_title);

      const title = `${base_title} ${counter}`;

      if (!(title in parsed_snapshot)) {
        throw new Error(`Unmatched snapshot title \`${title}\``);
      }

      const snapshot_value = parsed_snapshot[title];

      if (typeof snapshot_value !== 'string') {
        throw new Error(
          `Snapshot value should be a string, but got ${JSON.stringify(
            snapshot_value,
          )}`,
        );
      }

      const snapshot_value_first_line = snapshot_value.split('\n')[0];
      source_line_contents[trigger.line] += ` -> ${snapshot_value_first_line}`;
    });

  return source_line_contents.join('\n');

  function get_increased_counter(title: string) {
    const counter = default_to<number>(counters[title], 0) + 1;
    counters[title] = counter;
    return counter;
  }
};

function get_snapshot_base_title(trigger: Trigger) {
  const title = get_formatted_description(trigger, false);
  return trigger.group !== undefined
    ? `${trigger.group.title} ${title}`
    : title;
}
