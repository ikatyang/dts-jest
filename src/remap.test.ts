import * as ts from 'typescript';
import { load_fixture } from './helpers/load-fixture';
import { remap } from './remap';
import { snapshot_assertion_message } from './utils/create-assertion-expression';

it('should remap correctly', () => {
  expect(get_remapped('example')).toMatchSnapshot();
});

it('should throw error if there is unmatched snapshot', () => {
  expect(() => get_remapped('unmatched')).toThrowErrorMatchingSnapshot();
});

it('should throw error if there is non-string value in snapshot value', () => {
  const source_content = `
    // @dts-jest:snap
    Math.max(1, 2, 3);
  `;
  const snapshot_content = {
    [`Math.max(1, 2, 3) ${snapshot_assertion_message} 1`]: 123,
  };
  expect(() =>
    remap(source_content, snapshot_content),
  ).toThrowErrorMatchingSnapshot();
});

function get_remapped(id: string) {
  return remap(
    load_fixture(`remap/${id}.ts`),
    load_fixture(`remap/__snapshots__/${id}.ts.snap`),
    { typescript: ts },
  );
}
