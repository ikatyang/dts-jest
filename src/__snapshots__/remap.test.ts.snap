// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should remap correctly 1`] = `
"// @dts-jest:snap -> (mocked max)
Math.max(3, 2, 1);

// @dts-jest:snap description-min -> (mocked min)
Math.min(3, 2, 1);

// @dts-jest:group A

// @dts-jest:snap -> (mocked abs)
Math.abs(3);

// @dts-jest:snap description-sin -> (mocked sin)
Math.sin(0);

// @dts-jest:group B

// @dts-jest:snap -> (mocked cos)
Math.cos(0);

// @dts-jest:snap -> (mocked tan 1)
Math.tan(0);

// @dts-jest:snap -> (mocked tan 2)
Math.tan(0);
"
`;

exports[`should throw error if there is non-string value in snapshot value 1`] = `"Snapshot value should be a string, but got 123"`;

exports[`should throw error if there is unmatched snapshot 1`] = `"Unmatched snapshot title \`Math.max(4, 5, 6) (type) should match snapshot 1\`"`;
